### PR TITLE
If set mysql configuration in playbook do not override default value …

### LINKED ISSUE #26919: mysql_db defaults overriding config_file
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME
`mysql_db`

##### ANSIBLE VERSION
```
$ ansible --version
ansible 2.3.0.0
  config file = /Users/dylan/projects/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.11 (default, Jun 17 2016, 20:01:51) [GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)]
```

##### CONFIGURATION

n/a

##### OS / ENVIRONMENT

n/a

##### SUMMARY

.my.cnf is being overridden by defaults set by the module. specifically if a host is specified in the .my.cnf and not in the task, it seems to be overridden with `localhost`

##### STEPS TO REPRODUCE

I'm trying to dump a database with the following task: 

```yaml
- name: Dump databases
    mysql_db:
      state: dump
      db: api_db
      config_file: /home/dylan/.my.cnf
      target: /tmp/api_db.sql.gz
```

The .my.cnf provides the mysql user, password, and remote host url. However, it appears to be overridden by the default value set by login_host. When I don't explicitly set login_host (even though it's defined properly in the .my.cnf), I get `Cannot dump database api_db - not found` however if I explicitly set it in the task, eg
```yaml
- name: Dump databases
    mysql_db:
      state: dump
      db: api_db
      login_host: ........rds.amazonaws.com
      config_file: /home/dylan/.my.cnf
      target: /tmp/api_db.sql.gz
```
It completes successfully.

I would expect that the config file would override the defaults of the module instead of the other way around. 

Let me know if you can confirm this issue, if others agree I'm happy to dig into the code and submit a PR.

### PR DESCRIPTION
**Pull request by kucuny**
…only for host, port, socket (#26919)

##### SUMMARY
Fixed #26919

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
mysql_db

##### ANSIBLE VERSION
```
ansible 2.6.0 (prevent-override-login-host-when-already-set-mycnf a3c2e26f66) last updated 2018/05/15 13:44:21 (GMT -400)
  config file = None
  configured module search path = ['/Users/kucuny/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/kucuny/works/osp/ansible/lib/ansible
  executable location = /Users/kucuny/works/osp/ansible/bin/ansible
  python version = 3.6.1 (default, Mar 23 2017, 16:49:06) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

##### ADDITIONAL INFORMATION

```
# Test playbook yaml
---
- name: Mysql
  hosts: localhost
  gather_facts: no
  tasks:
    - name: dump from mysql
      mysql_db:
        name: <DATABASE_NAME>
        state: dump
        login_user: root
        login_password: root
        config_file: ~/my.cnf
        target: /tmp/testdump.sql.gz
```

```
# Test my.cnf
[client]
port        = 3306
host        = 127.0.0.1
```

```
ansible-playbook -v test.yml -i hosts
No config file found; using defaults

PLAY [Mysql] *********************************************************************************************************

TASK [dump from mysql] ***********************************************************************************************
changed: [localhost] => {"changed": true, "db": "<DATABASE_NAME>", "msg": ""}

PLAY RECAP ***********************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0
```
```
 ll /tmp/testdump.sql.gz
-rw-r--r--  1 kucuny  wheel  20 May 15 13:41 [1]  /tmp/testdump.sql.gz
```

### CODE REVIEW

File: lib/ansible/module_utils/mysql.py [hunk 1]
```diff
@@ -28,13 +28,34 @@
...
 except ImportError:
     mysqldb_found = False
 
+if PY2:
+    try:
+        import configparser
+    except ImportError:
+        import ConfigParser as configparser
+else:
+    import configparser
```
Comment (abadger) [inline lib/ansible/module_utils/mysql.py]: This can be directly imported from six:
``` python
from ansible.module_utils.six.moves import configparser
```

Comment (kucuny) [reply lib/ansible/module_utils/mysql.py]: @abadger I changed my code. Thanks for great review! :)

File: lib/ansible/module_utils/mysql.py [hunk 2]
```diff
@@ -36,13 +37,29 @@
...
     config = {}
 
     if ssl_ca is not None or ssl_key is not None or ssl_cert is not None:
         config['ssl'] = {}
 
+    cp = _parse_from_mysql_config_file(config_file)
+    if cp and cp.has_section('client'):
+        module.params['login_unix_socket'] = cp.get('client', 'socket', fallback=module.params['login_unix_socket'])
+        module.params['login_host'] = cp.get('client', 'host', fallback=module.params['login_host'])
+        module.params['login_port'] = cp.getint('client', 'port', fallback=module.params['login_port'])
```
Comment (abadger) [inline lib/ansible/module_utils/mysql.py]: This looks like it is unfortunately not quite this simple.  We want the following fallbacks:
* If the user specifies a value explicitly in the playbook, use that
* If the config file has it, then use that.
* If both of those do not have a value, then use the module's defaults.

This likely needs changes both here and in the mysql modules themselves.  The change here will look something like this (repeat for each of the values we're setting):
``` python
if cp an cp.has_section('client'):
    if module.params['login_host'] is None:
        module.params['login_host'] = cp.getint('client', 'host', fallback='localhost')
    [....]
```

Inside of the mysql modules we'll have to change teh module's argument_spec.  For instance, mysql_db.py has this:
``` python
        argument_spec=dict(
            login_user=dict(default=None),
            login_password=dict(default=None, no_log=True),
            login_host=dict(default="localhost"),
            [...]
```

We would need to change it to this:
``` python
        argument_spec=dict(
            login_user=dict(default=None),
            login_password=dict(default=None, no_log=True),
            login_host=dict(default=None),
            [...]
```

Comment (kucuny) [reply lib/ansible/module_utils/mysql.py]: @abadger Okay! I remove `login_host` default value and set 'socket', 'host' when these value is `None`. Thanks :)

**Comment by mattclay** [comment]
In addition to the sanity test failures the integration tests are failing:

https://app.shippable.com/github/ansible/ansible/runs/65942/25/tests

**Comment by Andersson007** [comment]
is it still relevant?
needs_info

**Pull request closed without merging**